sv_pipeline/load_data.py

```diff
@@ -0,0 +1,392 @@
+        header = f.readline()
+        header_indices = {col: i for i, col in enumerate(header.split())}
+        missing_cols = [col for col in columns or COLUMNS if col not in header_indices]
+        if missing_cols:
+            raise Exception('Missing expected columns: {}'.format(', '.join(missing_cols)))
+        if out_file:
+            out_file.write(header)
+
+        for line in tqdm(f, unit=' rows'):
+            row = line.split()
+            parsed = parse_row(row, header_indices)
+            if parsed and out_file:
+                out_file.write(line)
+
+    if out_file:
+        out_file.close()
+
+
+def subset_and_group_svs(input_dataset, sample_subset, sample_remap, ignore_missing_samples, write_subsetted_bed=False):
+    parsed_svs_by_name = {}
+    found_samples = set()
+    skipped_samples = set()
+    out_file_path = None
+    if write_subsetted_bed:
+        file_path = input_dataset.split('/')
+        file_path[-1] = 'subset_{}'.format(file_path[-1])
+        out_file_path ='/'.join(file_path)
+
+    def _parse_row(row, header_indices):
+        sample_id = get_field_val(row, SAMPLE_COL, header_indices)
+        if sample_remap and sample_id in sample_remap:
+            sample_id = sample_remap[sample_id]
+        if sample_subset is None or sample_id in sample_subset:
+            parse_sv_row(row, parsed_svs_by_name, header_indices, sample_id)
+            found_samples.add(sample_id)
+            return True
+        else:
+            skipped_samples.add(sample_id)
+            return False
+
+    load_file(input_dataset, _parse_row, out_file_path=out_file_path)
+
+    print('Found {} sample ids'.format(len(found_samples)))
+    if sample_subset:
+        if len(found_samples) != len(sample_subset):
+            missed_samples = sample_subset - found_samples
+            missing_sample_error = 'Missing the following {} samples:\n{}'.format(
+                len(missed_samples), ', '.join(sorted(missed_samples))
+            )
+            if ignore_missing_samples:
+                print(missing_sample_error)
+            else:
+                print('Samples in callset:\n{}'.format(', '.join(sorted(skipped_samples))))
+                raise Exception(missing_sample_error)
+
+    return parsed_svs_by_name
+
+
+def add_in_silico(svs_by_variant_id, file_path):
+    def _parse_row(row, header_indices):
+        variant_id = get_variant_id(row, header_indices)
+        if variant_id in svs_by_variant_id:
+            svs_by_variant_id[variant_id].update(get_parsed_column_values(row, header_indices, [IN_SILICO_COL]))
+
+    load_file(file_path, _parse_row, columns=IN_SILICO_COLS)
+
+
+def format_sv(sv):
+    sv[TRANSCRIPTS_FIELD] = [{'gene_id': gene} for gene in sv[GENES_FIELD]]
+    sv['transcriptConsequenceTerms'] = [sv[CALL_FIELD]]
+    if sv[SF_FIELD]:
+        sv['sn'] = int(sv[SC_FIELD] / sv[SF_FIELD])
+    sv['pos'] = sv[START_COL]
+    sv['xpos'] = CHROM_TO_XPOS_OFFSET[sv[CHROM_FIELD]] + sv[START_COL]
```

Comment:
what is the purpose of this `CHROM_TO_XPOS_OFFSET`? is this a formatting change for export to elasticsearch?

Reply:
`xpos` is a representation of position by a single number instead of as a chrom-pos pair and seqr heavily relies on it but its doesn't exist in this raw data so we have to compute it. So for example `chr1:248367227` has an `xpos` of 1248367227. `CHROM_TO_XPOS_OFFSET` is how we figure out how to represent the chromosome - for the numbered chromosomes its just `chr * 1,000,000,000` (so 1000000000, 2000000000, 3000000000, etc) but we can't use that formula here because X and Y are not actually numbers but are real chromosomes, so we precompute the offsets by chromosome in a way that more easily handles those edge cases

Reply:
ooh I see, thanks for explaining that!